### PR TITLE
fix(plugin-catalog): Fix TechDocsReaderPageContext to handle case sensitive entityRefs

### PR DESCRIPTION
**Pull request by jakecrews74**
## Hey, I just made a Pull Request!

I am picking this PR up from @drankou, see [here](https://github.com/backstage/backstage/pull/11527).

This PR ensures the URL path for techdocs is always in lowercase. There are no search results on the techdocs page when the entity's kind starts with an uppercase in the URL.

Related Discord message: https://discord.com/channels/687207715902193673/770283289327566848/961235926242193418

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you! 👍

https://github.com/backstage/backstage/pull/11527#pullrequestreview-995024175 still applies tbh. Really don't want to have to add this workaround in any code that links to TechDocs, but rather fix it in TechDocs itself.

**Comment by jakecrews74** [comment]
@Rugvip To make sure I am understanding this correctly, would you be ok with the approach of me moving where we are using `toLowercaseEntityRefMaybe` to roughly around [here](https://github.com/backstage/backstage/blob/ee60ad9ea48150ddf013f12c1afc53e213bd0420/plugins/techdocs/src/reader/components/TechDocsReaderPage/TechDocsReaderPage.tsx#L86-L107)? Which would avoid users having to do this check in individual components using the context.

**Comment by Rugvip** [comment]
@jakecrews74 kind of. It's a bit too late at that point as the URL will already be lowercased, but what we should be able to do is to check whether the casing is correct or not. See https://github.com/backstage/backstage/pull/11527#discussion_r894358709 for some more thoughts, but I will leave it to @backstage/techdocs-core to iron out the details.

**Comment by jakecrews74** [comment]
@Rugvip @iamEAP  I'm ready for review on this again. I interpreted @iamEAP's comment when he said

> Ref gets passed down further via context, but we also [read entity metadata here](https://github.com/backstage/backstage/blob/ee60ad9ea48150ddf013f12c1afc53e213bd0420/plugins/techdocs-react/src/context.tsx#L110-L121) and could maybe look to pass down a modified entityRef if case sensitivity is needed.

as moving the case sensitivity check/remediation issue into the techdocs-react context like suggested. This avoids people having to be careful with the case sensitivity of their navigation links since now the context is being passed entityRef from the URL params already remediated.

**Comment by jakecrews74** [comment]
@Rugvip @iamEAP Can I get a review on this?

**Comment by benjdlambert** [comment]
@rugvip is out on vacations right now, so I'm kind of filling in the slack. Sorry I missed this one. I think this seems fine, just need a review from @backstage/techdocs-core here would be great too, just so that I haven't missed anything.

**Comment by camilaibs** [comment]
> @Rugvip is out on vacations right now, so I'm kind of filling in the slack. Sorry I missed this one. I think this seems fine, just need a review from @backstage/techdocs-core here would be great too, just so that I haven't missed anything.

Starting to review it!

File: plugins/techdocs-react/src/context.test.tsx
Comment (camilaibs) [inline plugins/techdocs-react/src/context.test.tsx:152]: I suggest a small refactoring in this test to avoid defining "wrapper" twice:

Have only one "wrapper" definition in the root describe block and your config API mock as a global variable containing a getOptionalBoolean that returns undefined by default.

`const configApiMock = { getOptionalBoolean: jest.fn().mockReturnValue(undefined) };`

So here in this test you can change the return value once (`configApiMock.getOptionalBoolean.mockReturnValueOnce(true)`) before rendering the hook instead of defining a wrapper again.

**Review by camilaibs — COMMENTED** [review]
Thanks for contributing, I just added a small suggestion in the context test, other than that it looks great!

File: plugins/techdocs-react/src/context.test.tsx
Comment (jakecrews74) [reply plugins/techdocs-react/src/context.test.tsx]: Thanks for the pointers, this should be fixed!

**Review by camilaibs — APPROVED** [review]

**Comment by jakecrews74** [comment]
@camilaibs Thank you! Will I need to wait for @Rugvip to get back so he can remove his blocker in order to merge this?

**Comment by camilaibs** [comment]
> @camilaibs Thank you! Will I need to wait for @Rugvip to get back so he can remove his blocker in order to merge this?

I think @benjdlambert can help with this...

**Review by benjdlambert — APPROVED** [review]
Let's go! :ship:

**Pull request merged**
